File: leetcode/5.最长回文子串.go
/*
 * @lc app=leetcode.cn id=5 lang=go
 *
 * [5] 最长回文子串
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
 func longestPalindrome(s string) string {
	if s == "" {
			return ""
	}
	start, end := 0, 0
	for i := 0; i < len(s); i++ {
			left1, right1 := expandAroundCenter(s, i, i)
			left2, right2 := expandAroundCenter(s, i, i + 1)
			if right1 - left1 > end - start {
					start, end = left1, right1
			}
			if right2 - left2 > end - start {
					start, end = left2, right2
			}
	}
	return s[start:end+1]
}

func expandAroundCenter(s string, left, right int) (int, int) {
	leng := len(s);

	for ; left >= 0 && right < leng && s[left] == s[right]; left, right = left-1 , right+1 { }

	return left + 1, right - 1
}
// @lc code=end

File: leetcode/questions/5.最长回文子串.js
/*
 * @lc app=leetcode.cn id=5 lang=javascript
 *
 * [5] 最长回文子串
 */

// @lc code=start
/**
 * @param {string} s
 * @return {string}
 */
var longestPalindrome = function (s) {
  const leng = s.length;
  if (leng === 1) {
    return 1;
  }
  let left = 0, right = 0;
  let leftIdx = 0, rightIdx = 1, maxLen = 1;
  for (let i = 1; i < leng; i++) {
    const center = s[i];
    left = i - 1, right = i + 1;
    leftIdx = i, rightIdx = i;
    while (left<=right && left>=0 && right<=leng) {
      if ( (center === s[left] && center === s[right]) || s[left] === s[right] ) {
        left--;
        leftIdx--;
        right++;
        rightIdx++;
      }
      break;
    }
    maxLen = rightIdx - leftIdx;
  }
  return s.slice(leftIdx, rightIdx);
};
// @lc code=end

